src/coreclr/vm/amd64/jithelpers_fast.S

```diff
@@ -529,3 +529,27 @@ LEAF_ENTRY JIT_DispatchIndirectCall, _TEST
         movabs r11, 0xCDCDCDCDCDCDCDCD
         rex64 jmp rax
 LEAF_END JIT_DispatchIndirectCall, _TEST
+
+// Calculates the offset of native thread local variable `t_ThreadStatics`. The offset has to be found at runtime
+// once linker does its relocation and fixup of thread locals. It retrieves the current `instruction pointer` so
+// the caller can walk through the instruction bytes to retrieve the offset embedded by the linker and calculate the
+// final offset that should be passed to __tls_get_addr() in order to calculate the address of `t_ThreadStatics` for
+// the current thread. Here, we have to call `__tls_get_addr()`, because if the linker tries to find the code pattern
+// of "lea t_ThreadStatics@TLSGD", followed by `call __tls_get_addr()`. Without adding the call, the linker complains.
+// We do not use the result we get from `__tls_get_addr()` call, but instead just return the `instructino pointer` value
+// that we read in the beginning.
+LEAF_ENTRY JIT_GetThreadStaticsBaseOffset, _TEXT
+        push rbx
+        push rdi
+        lea rbx, [rip]                        // return the current instruction pointer
+        data16
+        lea rdi, t_ThreadStatics@TLSGD[rip]   // instruction where offset is embedded by the linker during compilation
+        data16
+        data16
+        rex64
+        call __tls_get_addr                   // dummy call to have linker see the code pattern to replace the offset
+        mov rax, rbx                          // return the instruction pointer
+        pop rdi
+        pop rbx
+        ret
```

Comment:
```suggestion
LEAF_ENTRY JIT_GetThreadStaticsBaseOffset, _TEXT
        data16
        lea rdi, t_ThreadStatics@TLSGD[rip]   // instruction where offset is embedded by the linker during compilation
        data16
        data16
        rex64
        call __tls_get_addr                   // dummy call to have linker see the code pattern to replace the offset
        int 3
```
Just an idea.  Now that it is in .S file, you do not need to call this helper to return the address of the interesting asm sequence. Instead, you can decode the interesting asm sequence directly.

Reply:
Good idea. It should work.